demos/smart_classroom_demo_gapi/src/custom_kerenels.cpp

```diff
@@ -0,0 +1,408 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+#include "kernel_packages.hpp"
+#include "custom_kernels.hpp"
+
+/** State parameters for RecognizeResultPostProc stateful kernel **/
+struct PostProcState {
+    PostProcState(const LoggerParams& params) : logger(DetectionsLogger(params.write_)) {}
+    DetectionsLogger logger;
+    int teacher_track_id = -1;
+    std::vector<std::map<int, int>> face_obj_id_to_action_maps;
+};
+
+/** State parameter for TopAction stateful kernel **/
+struct TopKState {
+    std::map<int, int> top_k_obj_ids;
+};
+
+/** State parameters for GetRecognitionResult stateful kernel **/
+struct TrackerState {
+    TrackerState(TrackerParams tracker_reid_params, TrackerParams tracker_action_params)
+        : tracker_reid(Tracker(tracker_reid_params)), tracker_action(Tracker(tracker_action_params)) {}
+    Tracker tracker_reid;
+    Tracker tracker_action;
+};
+
+/** Colors for labels and ROI **/
+const cv::Scalar red_color   = CV_RGB(255, 0,   0);
+const cv::Scalar white_color = CV_RGB(255, 255, 255);
+
+GAPI_OCV_KERNEL(OCVFaceDetectorPostProc, custom::FaceDetectorPostProc) {
+    static void run(const cv::Mat& in,
+                    const std::vector<cv::Rect>& rois,
+                    const detection::FaceDetectionKernelInput &face_inp,
+                    std::vector<cv::Rect>& out_rects) {
+        face_inp.ptr->truncateRois(in, rois, out_rects);
+    }
+};
+
+GAPI_OCV_KERNEL(OCVGetRectFromImage, custom::GetRectFromImage) {
+    static void run(const cv::Mat& in_image,
+                    std::vector<cv::Rect>& out_rects) {
+        out_rects.emplace_back(cv::Rect(0, 0, in_image.cols, in_image.rows));
+    }
+};
+
+GAPI_OCV_KERNEL(OCVAlignFacesForReidentification,
+                custom::AlignFacesForReidentification) {
+    static void run(const cv::Mat& in,
+                    const std::vector<cv::Mat>& landmarks,
+                    const std::vector<cv::Rect>& face_rois,
+                          std::vector<cv::Mat>& out_images) {
+        cv::Mat out_image = in.clone();
```

Comment:
Do you really need this clone?
Does AlignFaces works really inplace or  reallocate its every face image?